### PR TITLE
More helpful error message for `ExtensionList.lookupSingleton`

### PR DESCRIPTION
**Pull request by jglick**
A periodic issue, most recently noticed going over something with @julieheard: if you run `JenkinsRule` tests using a special test runner from an IDE which does not autodiscover annotation processors, things will be badly broken, and a typical symptom is this otherwise fairly opaque error. If in doubt, use Maven to run tests (and invoke the `clean` phase to be sure).

### Testing done

None, my IDE runs Maven (well, `mvnd`) to begin with! **Edit**: https://github.com/jenkinsci/jenkins/pull/8646#issuecomment-1781448588

### Proposed changelog entries

- N/A

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Review by Vlatombe — APPROVED** [review]

**Review by julieheard — APPROVED** [review]

**Comment by jglick** [comment]
Found a way to test, sort of. After

```bash
mvnd -am -pl war -Pquick-build install
cd ../branch-api-plugin
```

and

```diff
diff --git src/main/java/jenkins/branch/WorkspaceLocatorImpl.java src/main/java/jenkins/branch/WorkspaceLocatorImpl.java
index 4515492..14d452b 100644
--- src/main/java/jenkins/branch/WorkspaceLocatorImpl.java
+++ src/main/java/jenkins/branch/WorkspaceLocatorImpl.java
@@ -82,7 +82,6 @@ import org.kohsuke.accmod.restrictions.NoExternalUse;
  * @see "JENKINS-2111"
  */
 @Restricted(NoExternalUse.class)
-@Extension(ordinal = -100)
 public class WorkspaceLocatorImpl extends WorkspaceLocator {
 
     private static final Logger LOGGER = Logger.getLogger(WorkspaceLocatorImpl.class.getName());
```

Before:

```console
$ mvn test -Dtest=WorkspaceLocatorImplTest\#locate
…
   4.380 [id=39]	INFO	jenkins.InitReactorRunner$1#onAttained: Completed initialization
   4.413 [id=20]	WARNING	jenkins.model.Jenkins#<init>: Exception in onOnline() for the computer listener class jenkins.branch.WorkspaceLocatorImpl$Collector on the built-in node
java.lang.IllegalStateException: Expected 1 instance of jenkins.branch.WorkspaceLocatorImpl but got 0
	at hudson.ExtensionList.lookupSingleton(ExtensionList.java:454)
	at jenkins.branch.WorkspaceLocatorImpl.lockFor(WorkspaceLocatorImpl.java:310)
	at jenkins.branch.WorkspaceLocatorImpl$Collector.onOnline(WorkspaceLocatorImpl.java:631)
	at jenkins.model.Jenkins.<init>(Jenkins.java:1026)
	at hudson.model.Hudson.<init>(Hudson.java:86)
	at org.jvnet.hudson.test.JenkinsRule.newHudson(JenkinsRule.java:697)
	at org.jvnet.hudson.test.JenkinsRule.before(JenkinsRule.java:406)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:601)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
…
ERROR] Failures: 
[ERROR]   WorkspaceLocatorImplTest.locate:157 expected:<…/branch-api-plugin/target/tmp/j h16047186359448965079/workspace/stuff_dev_flow> but was:<…/branch-api-plugin/target/tmp/j h16047186359448965079/workspace/stuff/dev%2Fflow>
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
…
```

After:

```console
$ mvn test -Dtest=WorkspaceLocatorImplTest\#locate -Djenkins.version=2.430-SNAPSHOT
…
java.lang.IllegalStateException: Found no instances of jenkins.branch.WorkspaceLocatorImpl registered (possible annotation processor issue); try using `mvn clean test -Dtest=…` rather than an IDE test runner
…
```

**Review by NotMyFault — APPROVED** [review]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
